File: github_sync/git_client.py
"""
Async Git operations client for branch state analysis.

Wraps git commands in async executors for non-blocking operation.
"""

import asyncio
import subprocess
from datetime import datetime
from pathlib import Path
from typing import Any

from github_sync.models import Branch, BranchStatus


class GitClientError(Exception):
    """Exception raised for git operation errors."""


class GitClient:
    """Async wrapper for git operations."""

    def __init__(self, repo_path: Path | str):
        """
        Initialize git client for a repository.

        Args:
            repo_path: Path to the git repository
        """
        self.repo_path = Path(repo_path)

    async def validate_repo(self) -> None:
        """Validate that the path is a valid git repository."""
        if not self.repo_path.exists():
            raise GitClientError(f"Repository path does not exist: {self.repo_path}")

        git_dir = self.repo_path / ".git"
        if not git_dir.exists():
            raise GitClientError(f"Not a git repository: {self.repo_path}")

    async def _run_command(
        self,
        cmd: list[str],
        timeout: int = 30,
        check: bool = True,
    ) -> tuple[int, str, str]:
        """
        Run a git command asynchronously.

        Args:
            cmd: Command and arguments
            timeout: Timeout in seconds
            check: Raise exception on non-zero exit code

        Returns:
            Tuple of (returncode, stdout, stderr)
        """
        loop = asyncio.get_event_loop()

        def run_sync() -> tuple[int, str, str]:
            try:
                result = subprocess.run(
                    cmd,
                    check=False,
                    cwd=self.repo_path,
                    capture_output=True,
                    text=True,
                    timeout=timeout,
                )
                return result.returncode, result.stdout, result.stderr
            except subprocess.TimeoutExpired:
                raise GitClientError(f"Command timed out: {' '.join(cmd)}")

        returncode, stdout, stderr = await loop.run_in_executor(None, run_sync)

        if check and returncode != 0:
            raise GitClientError(
                f"Command failed: {' '.join(cmd)}\nExit code: {returncode}\nStderr: {stderr}"
            )

        return returncode, stdout, stderr

    async def fetch_all(self, prune: bool = True) -> None:
        """
        Fetch all remotes.

        Args:
            prune: Whether to prune deleted remote branches
        """
        cmd = ["git", "fetch", "--all"]
        if prune:
            cmd.append("--prune")

        await self._run_command(cmd, timeout=60)

    async def get_local_branches(self) -> list[Branch]:
        """Get all local branches with their information."""
        cmd = [
            "git",
            "for-each-ref",
            "--format=%(refname:short)|%(objectname)|%(upstream:short)|%(committerdate:iso8601)|%(authorname)|%(subject)",
            "refs/heads/",
        ]

        _, stdout, _ = await self._run_command(cmd)

        branches = []
        for line in stdout.strip().split("\n"):
            if not line:
                continue

            parts = line.split("|")
            if len(parts) < 6:
                continue

            name, sha, upstream, date_str, author, message = parts
            upstream_value: str | None = upstream if upstream else None

            # Parse date
            try:
                commit_date = datetime.fromisoformat(date_str.replace(" ", "T"))
            except (ValueError, AttributeError):
                commit_date = None

            branch = Branch(
                name=name,
                sha=sha,
                is_local=True,
                is_remote=False,
                upstream=upstream_value,
                last_commit_date=commit_date,
                last_commit_author=author,
                last_commit_message=message,
            )

            branches.append(branch)

        return branches

    async def get_remote_branches(self, remote: str = "origin") -> list[Branch]:
        """Get all remote branches."""
        cmd = [
            "git",
            "for-each-ref",
            "--format=%(refname:short)|%(objectname)|%(committerdate:iso8601)|%(authorname)|%(subject)",
            f"refs/remotes/{remote}/",
        ]

        _, stdout, _ = await self._run_command(cmd)

        branches = []
        for line in stdout.strip().split("\n"):
            if not line:
                continue

            parts = line.split("|")
            if len(parts) < 5:
                continue

            full_name, sha, date_str, author, message = parts

            # Remove remote prefix from name
            name = full_name.replace(f"{remote}/", "")

            # Skip HEAD reference
            if name == "HEAD":
                continue

            # Parse date
            try:
                commit_date = datetime.fromisoformat(date_str.replace(" ", "T"))
            except (ValueError, AttributeError):
                commit_date = None

            branch = Branch(
                name=name,
                sha=sha,
                is_local=False,
                is_remote=True,
                last_commit_date=commit_date,
                last_commit_author=author,
                last_commit_message=message,
            )

            branches.append(branch)

        return branches

    async def compute_branch_status(
        self,
        local_branch: Branch,
        remote_branch: Branch | None,
        remote: str = "origin",
    ) -> tuple[BranchStatus, int, int]:
        """
        Compute the status of a local branch relative to its remote.

        Args:
            local_branch: Local branch object
            remote_branch: Remote branch object (or None if no remote)
            remote: Remote name

        Returns:
            Tuple of (status, ahead_count, behind_count)
        """
        if remote_branch is None:
            return BranchStatus.LOCAL_ONLY, 0, 0

        # Compare SHAs
        if local_branch.sha == remote_branch.sha:
            return BranchStatus.UP_TO_DATE, 0, 0

        # Get ahead/behind counts
        try:
            cmd = [
                "git",
                "rev-list",
                "--left-right",
                "--count",
                f"{local_branch.name}...{remote}/{remote_branch.name}",
            ]

            returncode, stdout, _ = await self._run_command(cmd, check=False)

            if returncode != 0:
                return BranchStatus.UNKNOWN, 0, 0

            parts = stdout.strip().split()
            if len(parts) != 2:
                return BranchStatus.UNKNOWN, 0, 0

            ahead = int(parts[0])
            behind = int(parts[1])

            if ahead > 0 and behind > 0:
                return BranchStatus.DIVERGED, ahead, behind
            if ahead > 0:
                return BranchStatus.AHEAD, ahead, behind
            if behind > 0:
                return BranchStatus.BEHIND, ahead, behind
            return BranchStatus.UP_TO_DATE, ahead, behind

        except (ValueError, GitClientError):
            return BranchStatus.UNKNOWN, 0, 0

    async def get_all_branches_with_status(
        self,
        remote: str = "origin",
    ) -> tuple[list[Branch], list[Branch]]:
        """
        Get all branches (local and remote) with computed status.

        Args:
            remote: Remote name

        Returns:
            Tuple of (local_branches, remote_branches) with status computed
        """
        # Fetch local and remote branches concurrently
        local_branches, remote_branches = await asyncio.gather(
            self.get_local_branches(),
            self.get_remote_branches(remote),
        )

        # Build lookup for remote branches
        remote_lookup = {branch.name: branch for branch in remote_branches}

        # Compute status for each local branch
        for local_branch in local_branches:
            remote_branch = remote_lookup.get(local_branch.name)

            if remote_branch:
                status, ahead, behind = await self.compute_branch_status(
                    local_branch,
                    remote_branch,
                    remote,
                )
                local_branch.status = status
                local_branch.ahead_by = ahead
                local_branch.behind_by = behind
            else:
                local_branch.status = BranchStatus.LOCAL_ONLY

        return local_branches, remote_branches

    async def get_default_branch(self) -> str:
        """Detect the default branch of the repository."""
        # Try to get from remote HEAD
        cmd = ["git", "symbolic-ref", "refs/remotes/origin/HEAD"]
        returncode, stdout, _ = await self._run_command(cmd, check=False)

        if returncode == 0 and stdout:
            branch = stdout.strip().split("/")[-1]
            if branch:
                return branch

        # Fallback: check common branch names
        for branch in ["main", "master", "develop"]:
            cmd = ["git", "rev-parse", "--verify", f"origin/{branch}"]
            returncode, _, _ = await self._run_command(cmd, check=False)
            if returncode == 0:
                return branch

        return "main"

    async def get_current_branch(self) -> str | None:
        """Get the currently checked out branch."""
        cmd = ["git", "rev-parse", "--abbrev-ref", "HEAD"]
        returncode, stdout, _ = await self._run_command(cmd, check=False)

        if returncode == 0 and stdout:
            branch = stdout.strip()
            if branch != "HEAD":
                return branch

        return None

    async def get_repository_info(self) -> dict[str, Any]:
        """Get general repository information."""
        info: dict[str, Any] = {
            "path": str(self.repo_path),
        }

        # Run these concurrently
        default_branch_task = self.get_default_branch()
        current_branch_task = self.get_current_branch()

        info["default_branch"], info["current_branch"] = await asyncio.gather(
            default_branch_task, current_branch_task
        )

        # Get remote URL
        cmd = ["git", "remote", "get-url", "origin"]
        returncode, stdout, _ = await self._run_command(cmd, check=False)
        if returncode == 0:
            info["remote_url"] = stdout.strip()

        return info

    async def get_repo_name(self) -> str | None:
        """Extract repository name from remote URL."""
        cmd = ["git", "remote", "get-url", "origin"]
        returncode, stdout, _ = await self._run_command(cmd, check=False)

        if returncode != 0:
            return None

        url = stdout.strip()

        if "github.com" not in url:
            return None

        if url.startswith("git@github.com:"):
            repo_part = url.replace("git@github.com:", "").replace(".git", "")
        elif "github.com/" in url:
            repo_part = url.split("github.com/")[-1].replace(".git", "")
        else:
            return None

        parts = repo_part.split("/")
        if len(parts) >= 2:
            return parts[1]  # Return just the repo name

        return None

    async def get_repo_full_name(self) -> str | None:
        """Extract owner/repo from remote URL."""
        cmd = ["git", "remote", "get-url", "origin"]
        returncode, stdout, _ = await self._run_command(cmd, check=False)

        if returncode != 0:
            return None

        url = stdout.strip()

        if "github.com" not in url:
            return None

        if url.startswith("git@github.com:"):
            return url.replace("git@github.com:", "").replace(".git", "")
        if "github.com/" in url:
            return url.split("github.com/")[-1].replace(".git", "")

        return None

File: packages/github-sync/src/git-client.ts
/**
 * Local git client for @merge-god/github-sync.
 *
 * Thin async wrapper around synchronous `git` subprocess calls. Ports the
 * branch-analysis logic from the root `git_ops.ts` (merge-god) onto the shared
 * `Branch` / `BranchStatus` models in `./models`. All filesystem/git access is
 * synchronous via `spawnSync`; methods are `async` only to satisfy the
 * forge/store async library API.
 */

import { existsSync } from "node:fs";
import path from "node:path";
import { spawnSync } from "node:child_process";
import { BranchStatus, createBranch, type Branch } from "./models";

/** Error raised by {@link GitClient}. */
export class GitClientError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "GitClientError";
  }
}

interface CommandResult {
  returncode: number;
  stdout: string;
  stderr: string;
}

function runCommand(
  cwd: string,
  cmd: string[],
  opts: { timeout?: number; check?: boolean } = {},
): CommandResult {
  const { timeout = 30, check = true } = opts;
  try {
    const result = spawnSync(cmd[0] ?? "", cmd.slice(1), {
      cwd,
      encoding: "utf8",
      timeout: timeout * 1000,
      maxBuffer: 10 * 1024 * 1024,
    });

    if (result.error) {
      if (result.signal === "SIGTERM") {
        throw new GitClientError(`Command timed out: ${cmd.join(" ")}`);
      }
      throw new GitClientError(`Command error: ${result.error.message}`);
    }

    const returncode = result.status ?? -1;
    if (check && returncode !== 0) {
      throw new GitClientError(
        `Command failed: ${cmd.join(" ")}\n` +
          `Exit code: ${returncode}\n` +
          `Stderr: ${result.stderr ?? ""}`,
      );
    }

    return {
      returncode,
      stdout: result.stdout ?? "",
      stderr: result.stderr ?? "",
    };
  } catch (e) {
    if (e instanceof GitClientError) throw e;
    throw new GitClientError(`Command error: ${e instanceof Error ? e.message : String(e)}`);
  }
}

function parseDate(s: string): Date | null {
  if (!s) return null;
  const normalized = s.replaceAll(" ", "T");
  const d = new Date(normalized);
  if (Number.isNaN(d.getTime())) return null;
  return d;
}

/**
 * Local-git client. Wraps `git` subprocess invocations for a single checkout.
 */
export class GitClient {
  readonly repoPath: string;

  /** @param repoPath - Path to the git working tree. */
  constructor(repoPath: string) {
    this.repoPath = path.resolve(repoPath);
    this.validateRepo();
  }

  private validateRepo(): void {
    if (!existsSync(this.repoPath)) {
      throw new GitClientError(`Repository path does not exist: ${this.repoPath}`);
    }
    const gitDir = path.join(this.repoPath, ".git");
    if (!existsSync(gitDir)) {
      throw new GitClientError(`Not a git repository: ${this.repoPath}`);
    }
  }

  private runCommand(
    cmd: string[],
    opts: { timeout?: number; check?: boolean } = {},
  ): CommandResult {
    return runCommand(this.repoPath, cmd, opts);
  }

  /** Fetch all remotes (with prune). */
  fetch(remote = "origin"): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        this.runCommand(["git", "fetch", remote, "--prune"], { timeout: 30 });
        resolve();
      } catch (e) {
        reject(e instanceof Error ? e : new GitClientError(String(e)));
      }
    });
  }

  /** List local branches (refs/heads). */
  getLocalBranches(): Promise<Branch[]> {
    return new Promise((resolve, reject) => {
      try {
        const cmd = [
          "git",
          "for-each-ref",
          "--format=%(refname:short)|%(objectname)|%(upstream:short)|%(committerdate:iso8601)|%(authorname)|%(subject)",
          "refs/heads/",
        ];
        const { stdout } = this.runCommand(cmd);

        const branches: Branch[] = [];
        for (const line of stdout.trim().split("\n")) {
          if (!line) continue;

          const parts = line.split("|");
          if (parts.length < 6) continue;

          const name = parts[0]!;
          const sha = parts[1]!;
          const upstream = parts[2]!;
          const dateStr = parts[3]!;
          const author = parts[4]!;
          const message = parts.slice(5).join("|");

          const commitDate = parseDate(dateStr);

          branches.push(
            createBranch({
              name,
              sha,
              is_local: true,
              is_remote: false,
              upstream: upstream || null,
              last_commit_date: commitDate,
              last_commit_author: author,
              last_commit_message: message,
            }),
          );
        }

        resolve(branches);
      } catch (e) {
        reject(e instanceof Error ? e : new GitClientError(String(e)));
      }
    });
  }

  /** List remote branches under `refs/remotes/<remote>/`. */
  getRemoteBranches(remote = "origin"): Promise<Branch[]> {
    return new Promise((resolve, reject) => {
      try {
        const cmd = [
          "git",
          "for-each-ref",
          "--format=%(refname:short)|%(objectname)|%(committerdate:iso8601)|%(authorname)|%(subject)",
          `refs/remotes/${remote}/`,
        ];
        const { stdout } = this.runCommand(cmd);

        const branches: Branch[] = [];
        for (const line of stdout.trim().split("\n")) {
          if (!line) continue;

          const parts = line.split("|");
          if (parts.length < 5) continue;

          const fullName = parts[0]!;
          const sha = parts[1]!;
          const dateStr = parts[2]!;
          const author = parts[3]!;
          const message = parts.slice(4).join("|");

          const name = fullName.replaceAll(`${remote}/`, "");
          if (name === "HEAD") continue;

          const commitDate = parseDate(dateStr);

          branches.push(
            createBranch({
              name,
              sha,
              is_local: false,
              is_remote: true,
              last_commit_date: commitDate,
              last_commit_author: author,
              last_commit_message: message,
            }),
          );
        }

        resolve(branches);
      } catch (e) {
        reject(e instanceof Error ? e : new GitClientError(String(e)));
      }
    });
  }

  /**
   * Compute the ahead/behind relationship between a local branch and its
   * remote counterpart.
   *
   * @returns `[status, ahead_by, behind_by]`.
   */
  computeBranchStatus(
    localBranch: Branch,
    remoteBranch: Branch | null,
    remoteName = "origin",
  ): [BranchStatus, number, number] {
    if (remoteBranch === null) {
      return [BranchStatus.LOCAL_ONLY, 0, 0];
    }

    if (localBranch.sha === remoteBranch.sha) {
      return [BranchStatus.UP_TO_DATE, 0, 0];
    }

    try {
      const cmd = [
        "git",
        "rev-list",
        "--left-right",
        "--count",
        `${localBranch.name}...${remoteName}/${remoteBranch.name}`,
      ];
      const { returncode, stdout } = this.runCommand(cmd, { check: false });

      if (returncode !== 0) {
        return [BranchStatus.UNKNOWN, 0, 0];
      }

      const parts = stdout.trim().split(/\s+/);
      if (parts.length !== 2) {
        return [BranchStatus.UNKNOWN, 0, 0];
      }

      const ahead = Number(parts[0]);
      const behind = Number(parts[1]);
      if (Number.isNaN(ahead) || Number.isNaN(behind)) {
        return [BranchStatus.UNKNOWN, 0, 0];
      }

      if (ahead > 0 && behind > 0) return [BranchStatus.DIVERGED, ahead, behind];
      if (ahead > 0) return [BranchStatus.AHEAD, ahead, behind];
      if (behind > 0) return [BranchStatus.BEHIND, ahead, behind];
      return [BranchStatus.UP_TO_DATE, ahead, behind];
    } catch {
      return [BranchStatus.UNKNOWN, 0, 0];
    }
  }

  /**
   * Get local + remote branches with ahead/behind status computed on every
   * local branch that has a remote counterpart.
   *
   * @returns `[localBranches, remoteBranches]`.
   */
  getAllBranchesWithStatus(remote = "origin"): Promise<[Branch[], Branch[]]> {
    return new Promise((resolve, reject) => {
      Promise.all([this.getLocalBranches(), this.getRemoteBranches(remote)])
        .then(([localBranches, remoteBranches]) => {
          const remoteLookup = new Map<string, Branch>();
          for (const branch of remoteBranches) remoteLookup.set(branch.name, branch);

          for (const localBranch of localBranches) {
            const remoteBranch = remoteLookup.get(localBranch.name) ?? null;

            if (remoteBranch) {
              const [status, ahead, behind] = this.computeBranchStatus(
                localBranch,
                remoteBranch,
                remote,
              );
              localBranch.status = status;
              localBranch.ahead_by = ahead;
              localBranch.behind_by = behind;
            } else {
              localBranch.status = BranchStatus.LOCAL_ONLY;
            }
          }

          resolve([localBranches, remoteBranches]);
        })
        .catch((e) => reject(e instanceof Error ? e : new GitClientError(String(e))));
    });
  }

  /** Resolve the repository's default branch (origin/HEAD with fallbacks). */
  getDefaultBranch(): Promise<string> {
    return new Promise((resolve, reject) => {
      try {
        const cmd = ["git", "symbolic-ref", "refs/remotes/origin/HEAD"];
        const { returncode, stdout } = this.runCommand(cmd, { check: false });

        if (returncode === 0 && stdout) {
          const parts = stdout.trim().split("/");
          const branch = parts[parts.length - 1];
          if (branch) {
            resolve(branch);
            return;
          }
        }

        for (const branch of ["main", "master", "develop"]) {
          const { returncode: rc } = this.runCommand(
            ["git", "rev-parse", "--verify", `origin/${branch}`],
            { check: false },
          );
          if (rc === 0) {
            resolve(branch);
            return;
          }
        }

        resolve("main");
      } catch (e) {
        reject(e instanceof Error ? e : new GitClientError(String(e)));
      }
    });
  }

  /** Current checked-out branch name, or null if detached. */
  getCurrentBranch(): Promise<string | null> {
    return new Promise((resolve, reject) => {
      try {
        const cmd = ["git", "rev-parse", "--abbrev-ref", "HEAD"];
        const { returncode, stdout } = this.runCommand(cmd, { check: false });

        if (returncode === 0 && stdout) {
          const branch = stdout.trim();
          if (branch !== "HEAD") {
            resolve(branch);
            return;
          }
        }

        resolve(null);
      } catch (e) {
        reject(e instanceof Error ? e : new GitClientError(String(e)));
      }
    });
  }

  /** Basic repo metadata: path, default/current branch, remote URL. */
  getRepositoryInfo(): Promise<Record<string, unknown>> {
    return new Promise((resolve, reject) => {
      try {
        const info: Record<string, unknown> = {
          path: this.repoPath,
        };

        const defaultCmd = ["git", "symbolic-ref", "refs/remotes/origin/HEAD"];
        const def = this.runCommand(defaultCmd, { check: false });
        if (def.returncode === 0 && def.stdout) {
          const parts = def.stdout.trim().split("/");
          const branch = parts[parts.length - 1];
          if (branch) info.default_branch = branch;
        }

        const curCmd = ["git", "rev-parse", "--abbrev-ref", "HEAD"];
        const cur = this.runCommand(curCmd, { check: false });
        if (cur.returncode === 0 && cur.stdout) {
          const branch = cur.stdout.trim();
          if (branch !== "HEAD") info.current_branch = branch;
        }

        const urlCmd = ["git", "remote", "get-url", "origin"];
        const url = this.runCommand(urlCmd, { check: false });
        if (url.returncode === 0) {
          info.remote_url = url.stdout.trim();
        }

        resolve(info);
      } catch (e) {
        reject(e instanceof Error ? e : new GitClientError(String(e)));
      }
    });
  }
}
